core/src/main/scala/com/softwaremill/react/kafka_api_opt3/Example.scala

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (C) 2014 - 2016 Softwaremill <http://softwaremill.com>
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package com.softwaremill.react.kafka_api_opt3
+
+import scala.concurrent.Future
+
+import akka.Done
+import com.softwaremill.react.kafka_api_producer.Producer
+import org.apache.kafka.clients.consumer.KafkaConsumer
+import org.apache.kafka.clients.producer.KafkaProducer
+import org.apache.kafka.clients.producer.ProducerRecord
+
+trait Example {
+  implicit val ec = scala.concurrent.ExecutionContext.Implicits.global
+
+  def consumerProvider: () => KafkaConsumer[Array[Byte], String] = ???
+
+  def producerProvider: () => KafkaProducer[Array[Byte], String] = ???
+
+  class DB {
+    def save(offset: Long, data: String): Future[Done] = ???
+  }
+}
+
+// Consume messages and store a representation, including offset, in DB
+object Example3 extends Example {
+  val db = new DB
+
+  Consumer.source(consumerProvider)
```

Comment:
> Why anything should be lost during normal shutdown?

Commits are lost, because the message has been processed, but system has been shut down before Kafka has been notified. This was a reply to "We can not stop it because we have to be able to handle their commit and we have no idea when all these messages processed.". We can change the semantics though that stop just waits until all currently emitted messages have been committed, then stops the Kafka consumer and then emits onComplete.

> Not emitting onComplete may dead lock your stream (with splitAfter for example). 
I don't see any deadlock here. Can you build an example that deadlocks? 

> Also I may never call commit at all (in case of external offset storing)

But then I don't understand your original question. What *should* it do if you never commit? I mean, forget streams for now and just explain to me what you think are the desired semantics in normal non-stream terms so we can be on the same page (I fear some misunderstanding here).